D. Sravan Kumar

ID: BAE20012

Email_ID: davuluri.kumar20ug@apu.edu.in




# Second Graded Assignment
## Programming and data analysis
### Instructor: Avinash M. Tripathi

#### Instructions
1. To be submitted before semester end (May 7th midnight).
2. Write your name, email Id in the first cell of your answer script.
3. Use code cells to write necessary code. Markdown cells for textual answers.
4. Try to use working examples to illustrate your point.
5. __You should submit your own work and not seek help from anyone including your classmates. Any violation of this condition will be treated as use of unfair means.__
7. Upload your answer under graded assignment 2. If there is some technical glitch, you can mail me to complete your submission and upload your script later.
8. __You can see your lecture notes and textbooks while solving problems. You are not allowed to browse internet for solutions.__
9. You should look up your the lecture notes (including the practice sessions and class assignments) for clues for solving the problems.

10. To the extent possible, write idiomatic Pandas code. That is, minimize the use of for loops and work with bulit-in Pandas functions/methods instead.
11. Column names in datasets are mostly self-explanatory. If you are not clear about it, please send a mail.


__Q1.__ Describe the data analysis workflow in the Python. List any three challenges you are likely to face in data analysis and how will you solve them using the tools learnt in Unit 4 (__20 Marks__)

### Q1
   1)The first thing in the data analysis work flow is to choose the data set in a excel or csv format.

   2)Importing the required libraries such as pands, numpy, seaborn etc, which are required to analyze data.

   3)Loading the data in the python environment. For this, if it is csv file, I will use pd.read_csv() function and if it is excel file I will use pd.read_excel() function. I am comfortable in these two file formats most. So, /i prefer to these only.

   4)Cheaking how data set look likes, for this I use dataset.head() function where dataset is a variable where we stored our dataframe. I will also inspect to look for total number of rows and columns in the dataframe using dataset.shape method.

   5) Next, step is to ask for data set description using dataset.describe() function. It will give descriptive statistics of all float and numeric columns in the dataset.

   6) Now, I will check the datatype of each column in the dataset using dataset.info() function.

   7)In the previous step if there are object colums which are not intended to be object colums, I will use data cleaning methods to clease the data to create the correct type of column.For data cleaning there are so many methods, I will prefer to use a dataset.replace() function by creating user defined function.

   8)After cleaning the data, the next task is to see if the dataset has all the columns which are required perform our data analysis. Sometime, here we have to sum one or two columns, sometimes we have to square the existing column and create a new column with it. Sometimes we have to create bins like this, It will vary for case to case.

   9)After performing all the above tasks our real data analysis will start. It requires specific question to work on. What we want to infer from the data and what methods do we have to use to get our required results.

   10)After having the specific question we have, we will again download libraries based on the task we have to do. For example, if we want to do some regression analysis, then we have to import statmodels.

   11)Sometimes, we have to visualize data, find correlations between variables to make conclusions. So, we have to do that.

   12)After doing all these we have to evaluate our results that we got after performing all the tasks.

   13) And finally, we have to make our results in a organised way such that we can present them and make other people understand them.

   ##Problems likely to face
   1)The major problem is that, we can't get the data in a required format with required variables, this is the problem I am facing to do the last question. And also finding the data sets is to also one of the though question. We have tackle this somehow outside python environment.

   2) Sometimes even though we got the desired data it will be very mess and uncleaned. For this I will use the data cleaning techniques in the python.

   3) And finally sometimes even after doing everything we don't know how to evaluate and present the results that we have. It is bigger than doing python, for this we need a lot of observational skills and analyzation skills. These should be learned by building perspectives and by taking help of instructors.





__Q2__.  UPassembly2002ID.csv dataset contains candidate-wise data for UP assembly elections in 2002. Use this dataset for answering the following questions.

a) Tablulate the seats won by each party. (__10 Marks__)


b) Effective Number of Candidates (ENC)--a measure of the political competition in a constituency--is defined by the following formula.
$$ENC = \frac{1}{\Sigma_i^n x_i^2}$$

Where $x_i$ is the vote share secured by the candidate $i$. Compute Effective Number of Candidates for every assembly constituency in the dataset. (__5 Marks__)

c) Let us assume that the Ranked variable is not included in the dataset. How can you create it using the Vote column? (__5 Marks__)



In [ ]:
import pandas as pd
#df1 = pd.read_csv('UPassembly2002ID.csv')
#df1.info()

In [ ]:
df1.head()

Constituency_No Constituency_Name              Candidate    Votes  \
0                1           SEOHARA             QUTUBUDEEN  37853.0   
1                1           SEOHARA  DR. VED PARKASH SINGH  35704.0   
2                1           SEOHARA            RANA PRATAP  33410.0   
3                1           SEOHARA           ANWAR JAMEEL  25934.0   
4                1           SEOHARA            BAHAJUDDEEN  14381.0   

   Valid_Votes Party  Ranked  Vote_Share  
0     163780.0   BSP     1.0      0.2311  
1     163780.0   BJP     2.0      0.2180  
2     163780.0   IND     3.0      0.2040  
3     163780.0   NCP     4.0      0.1583  
4     163780.0    SP     5.0      0.0878

In [ ]:
# Question 2a
win = df1.loc[df1['Ranked']==1.0]
print(win.groupby('Party').size())
print('If a party not appear here, it got 0 seats')

Party
ABHM        1
ABLTC       2
AD          3
BJP        88
BSP        99
CPM         2
INC        25
IND        16
JD(U)       2
JP          1
LJNSP       1
NLP         1
RLD        14
RPD         1
RTKP        4
SJP(R)      1
SP        142
dtype: int64
If a party not appear here, it got 0 seats


In [ ]:
def vote_share_square(x):
  return x**2

In [ ]:
df1['vote_share_sqr'] = df1['Vote_Share'].apply(vote_share_square)

In [ ]:
df1.head(15)

Constituency_No Constituency_Name              Candidate    Votes  \
0                 1           SEOHARA             QUTUBUDEEN  37853.0   
1                 1           SEOHARA  DR. VED PARKASH SINGH  35704.0   
2                 1           SEOHARA            RANA PRATAP  33410.0   
3                 1           SEOHARA           ANWAR JAMEEL  25934.0   
4                 1           SEOHARA            BAHAJUDDEEN  14381.0   
5                 1           SEOHARA           CHHOTE SINGH   6430.0   
6                 1           SEOHARA          AVANISH KUMAR   6302.0   
7                 1           SEOHARA        RAGHUVEER SINGH    710.0   
8                 1           SEOHARA             RAJESHWARI    688.0   
9                 1           SEOHARA            NANAK SINGH    685.0   
10                1           SEOHARA           BADAR PARWEZ    581.0   
11                1           SEOHARA              TEJ SINGH    437.0   
12                1           SEOHARA        DR. CHANDRAWATI    354.0   
13                1           SEOHARA             RAM PRASAD    311.0   
14                2           DHAMPUR             MOOL CHAND  56597.0   

    Valid_Votes Party  Ranked  Vote_Share  vote_share_sqr  
0      163780.0   BSP     1.0      0.2311        0.053407  
1      163780.0   BJP     2.0      0.2180        0.047524  
2      163780.0   IND     3.0      0.2040        0.041616  
3      163780.0   NCP     4.0      0.1583        0.025059  
4      163780.0    SP     5.0      0.0878        0.007709  
5      163780.0   IND     6.0      0.0393        0.001544  
6      163780.0   INC     7.0      0.0385        0.001482  
7      163780.0   IND     8.0      0.0043        0.000018  
8      163780.0  INLD     9.0      0.0042        0.000018  
9      163780.0   RPI    10.0      0.0042        0.000018  
10     163780.0   IND    11.0      0.0035        0.000012  
11     163780.0   IND    12.0      0.0027        0.000007  
12     163780.0  RTKP    13.0      0.0022        0.000005  
13     163780.0  ABHM    14.0      0.0019        0.000004  
14     154501.0    SP     1.0      0.3663        0.134176

In [ ]:
df2 =pd.pivot_table(data = df1,
index= ['Constituency_Name'],
values = 'vote_share_sqr',
aggfunc= 'sum')

In [ ]:
df2

vote_share_sqr
Constituency_Name                  
AFZALGARH                  0.186062
AGOTA                      0.269156
AGRA CANTONMENT            0.282115
AGRA EAST                  0.255068
AGRA WEST                  0.292164
...                             ...
VARANASI CANTONMENT        0.200072
VARANASI NORTH             0.400459
VARANASI SOUTH             0.373235
ZAHOORABAD                 0.284261
ZAMANIA                    0.241366

[399 rows x 1 columns]

In [ ]:
def ENC(x):
  return (1/x)

In [ ]:
df2['ENC'] = df2['vote_share_sqr'].apply(ENC)

In [ ]:
# Question 2b final solution. Here I can use drop() to drop vote_share_sqr column to just show constituency anf ENC value but I am showing both.
df2

vote_share_sqr       ENC
Constituency_Name                            
AFZALGARH                  0.186062  5.374557
AGOTA                      0.269156  3.715324
AGRA CANTONMENT            0.282115  3.544656
AGRA EAST                  0.255068  3.920516
AGRA WEST                  0.292164  3.422736
...                             ...       ...
VARANASI CANTONMENT        0.200072  4.998199
VARANASI NORTH             0.400459  2.497137
VARANASI SOUTH             0.373235  2.679277
ZAHOORABAD                 0.284261  3.517890
ZAMANIA                    0.241366  4.143089

[399 rows x 2 columns]

In [ ]:
#Question 2c here Ranked1 corresponding to Ranked.
df1['Ranked1'] = df1.groupby('Constituency_Name')['Votes'].rank(ascending = False)



In [ ]:
df1

Constituency_No Constituency_Name                  Candidate    Votes  \
0                   1           SEOHARA                 QUTUBUDEEN  37853.0   
1                   1           SEOHARA      DR. VED PARKASH SINGH  35704.0   
2                   1           SEOHARA                RANA PRATAP  33410.0   
3                   1           SEOHARA               ANWAR JAMEEL  25934.0   
4                   1           SEOHARA                BAHAJUDDEEN  14381.0   
...               ...               ...                        ...      ...   
5546              403      MUZAFFARABAD         JAGDISH SINGH RANA  51914.0   
5547              403      MUZAFFARABAD         RAOMOHD NEAIM KHAN  43381.0   
5548              403      MUZAFFARABAD   THAKUR CHANDER PAL SINGH  24786.0   
5549              403      MUZAFFARABAD  JAMSHEDLI URF JAMSHED ALI   3569.0   
5550              403      MUZAFFARABAD    DR. ARJUN SINGH KASHYAP   3236.0   

      Valid_Votes Party  Ranked  Vote_Share  vote_share_sqr  Ranked1  
0        163780.0   BSP     1.0      0.2311        0.053407      1.0  
1        163780.0   BJP     2.0      0.2180        0.047524      2.0  
2        163780.0   IND     3.0      0.2040        0.041616      3.0  
3        163780.0   NCP     4.0      0.1583        0.025059      4.0  
4        163780.0    SP     5.0      0.0878        0.007709      5.0  
...           ...   ...     ...         ...             ...      ...  
5546     126886.0    SP     1.0      0.4091        0.167363      1.0  
5547     126886.0   BSP     2.0      0.3419        0.116896      2.0  
5548     126886.0   BJP     3.0      0.1953        0.038142      3.0  
5549     126886.0    JP     4.0      0.0281        0.000790      4.0  
5550     126886.0  RTKP     5.0      0.0255        0.000650      5.0  

[5551 rows x 10 columns]

__Q3__.Use the dataset wage.csv for answering the following questions.

a) Regress wage on education. With or without intercepts. Interpret coefficients, t-statistic, p-values, R-squared. (__10 Marks__)

b)  Regress wage on education with experience (expr) being a control. Interpret coefficients, t-statistic, p-values, R-squared. (__10 Marks__)

In [ ]:
wage = pd.read_csv('wage.csv')
wage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 24 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   wage      526 non-null    float64
 1   educ      526 non-null    int64  
 2   exper     526 non-null    int64  
 3   tenure    526 non-null    int64  
 4   nonwhite  526 non-null    int64  
 5   female    526 non-null    int64  
 6   married   526 non-null    int64  
 7   numdep    526 non-null    int64  
 8   smsa      526 non-null    int64  
 9   northcen  526 non-null    int64  
 10  south     526 non-null    int64  
 11  west      526 non-null    int64  
 12  construc  526 non-null    int64  
 13  ndurman   526 non-null    int64  
 14  trcommpu  526 non-null    int64  
 15  trade     526 non-null    int64  
 16  services  526 non-null    int64  
 17  profserv  526 non-null    int64  
 18  profocc   526 non-null    int64  
 19  clerocc   526 non-null    int64  
 20  servocc   526 non-null    int64 

In [ ]:
wage.head()

wage  educ  exper  tenure  nonwhite  female  married  numdep  smsa  \
0  3.10    11      2       0         0       1        0       2     1   
1  3.24    12     22       2         0       1        1       3     1   
2  3.00    11      2       0         0       0        0       2     0   
3  6.00     8     44      28         0       0        1       0     1   
4  5.30    12      7       2         0       0        1       1     0   

   northcen  ...  trcommpu  trade  services  profserv  profocc  clerocc  \
0         0  ...         0      0         0         0        0        0   
1         0  ...         0      0         1         0        0        0   
2         0  ...         0      1         0         0        0        0   
3         0  ...         0      0         0         0        0        1   
4         0  ...         0      0         0         0        0        0   

   servocc     lwage  expersq  tenursq  
0        0  1.131402        4        0  
1        1  1.175573      484        4  
2        0  1.098612        4        0  
3        0  1.791759     1936      784  
4        0  1.667707       49        4  

[5 rows x 24 columns]

In [ ]:
#Solving for 3a
from statsmodels.formula.api import ols
ols('wage~educ', data = wage).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   wage   R-squared:                       0.165
Model:                            OLS   Adj. R-squared:                  0.163
Method:                 Least Squares   F-statistic:                     103.4
Date:                Sat, 07 May 2022   Prob (F-statistic):           2.78e-22
Time:                        13:59:27   Log-Likelihood:                -1385.7
No. Observations:                 526   AIC:                             2775.
Df Residuals:                     524   BIC:                             2784.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.9049      0.685     -1.321      0.187      -2.250       0.441
educ           0.5414      0.053     10.167      0.000       0.437       0.646
==============================================================================
Omnibus:                      212.554   Durbin-Watson:                   1.824
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              807.843
Skew:                           1.861   Prob(JB):                    3.79e-176
Kurtosis:                       7.797   Cond. No.                         60.2
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Answer for 3a
coefficient of education = 0.5414, it means as education increased by 1 unit, the prediced wage is increased by 0.5414 units.Here, I write unit as unit because there is no description of actual units in the problem.

t-statistic = 10.167 (we can get this by deviding co-efficient value by standard error)
p-value = 0.000 (We can get this by looking the   corresponding p-value of t-statistic we get in standard normal table)

Here, t-statistic is greater than 1.96 which means the co-efficient of education is statistically significantly different from zero.

R_squared = 0.165



In [ ]:
# Solving for 3b
ols('wage~educ+exper', data = wage).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   wage   R-squared:                       0.225
Model:                            OLS   Adj. R-squared:                  0.222
Method:                 Least Squares   F-statistic:                     75.99
Date:                Sat, 07 May 2022   Prob (F-statistic):           1.07e-29
Time:                        13:59:34   Log-Likelihood:                -1366.0
No. Observations:                 526   AIC:                             2738.
Df Residuals:                     523   BIC:                             2751.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -3.3905      0.767     -4.423      0.000      -4.896      -1.885
educ           0.6443      0.054     11.974      0.000       0.539       0.750
exper          0.0701      0.011      6.385      0.000       0.049       0.092
==============================================================================
Omnibus:                      196.872   Durbin-Watson:                   1.820
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              727.160
Skew:                           1.715   Prob(JB):                    1.26e-158
Kurtosis:                       7.627   Cond. No.                         130.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Answer for 3b
Education co-efficient = 0.6443

t-statistic of education = 11.974 (It can be calculated in same method as mentioned above)

p-value = 0.000 (It can be calculated in same method as mentioned above)


co-efficient of experience = 0.011

t-statistic = 6.385

p-value = 0.00

R-squared = 0.225

This regression is preferred more than the previous one because of higher R-squared value and both Education co-efficient and Experience co-efficient are statistically significant.Both t-statistics are more than 1.96.

__Q4__. Read the data contained in clean.csv and clean it.(__20 Marks__)

In [ ]:
clean = pd.read_csv('clean.csv')
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   X       1000 non-null   object
dtypes: object(1)
memory usage: 7.9+ KB


In [ ]:
def cleaner_data_frame(value):
  value = value.replace(',','')
  value = value.replace('%','')
  value = value.replace('-','nan')
  return float(value)

In [ ]:
clean['X']=clean['X'].apply(cleaner_data_frame)
clean

X
0    73.0
1    68.0
2    89.0
3    46.0
4    95.0
..    ...
995  44.0
996   3.0
997  80.0
998  75.0
999   NaN

[1000 rows x 1 columns]

__Q5__. State a problem in your chosen area of study and solve it using what you have learnt in unit 4. The more challenging and realistic your problem, the better.  (__20 Marks__)

In [ ]:
# My chosen area of study is my state (Andhra Pradesh) 2019 elections.
# My first question is to see if NOTA got first place any where.
#second question is to find gategory distribution of winning candidates and gender distribution of winning candidates
#Third question is to find my favourite party category wise candidates and category wise winning members.
#And finally to see distribution of young candidates that is age less than 30 across parties.
import pandas as pd
data = pd.read_csv("Election of Andhra proper.csv")

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2297 entries, 0 to 2296
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0    AC NO.         2297 non-null   object 
 1   Constituency    2293 non-null   object 
 2   Candidate       2293 non-null   object 
 3   Gender          2118 non-null   object 
 4   Age             2118 non-null   float64
 5   Category        2118 non-null   object 
 6   Party           2293 non-null   object 
 7   General         2297 non-null   int64  
 8   Postal          1333 non-null   float64
 9   Total           2297 non-null   int64  
 10  % Votes Polled  2297 non-null   float64
dtypes: float64(3), int64(2), object(6)
memory usage: 197.5+ KB


In [ ]:
data.shape

(2297, 11)

In [ ]:
data['Ranked'] = data.groupby('Constituency')['Total'].rank(ascending = False)

In [ ]:
data.head(20)

AC NO. Constituency                        Candidate  Gender   Age  \
0        1  Ichchapuram                 1 ASHOK BENDALAM    MALE  37.0   
1        1  Ichchapuram                  2 PIRIYA SAIRAJ    MALE  42.0   
2        1  Ichchapuram                    3 DASARI RAJU    MALE  42.0   
3        1  Ichchapuram                           4 NOTA     NaN   NaN   
4        1  Ichchapuram               5 ESWARA RAO KOLLI    MALE  50.0   
5        1  Ichchapuram  6 JANNALA SURYAVARA PRASADA RAO    MALE  67.0   
6        1  Ichchapuram              7 GANAPA VANAJAKSHI  FEMALE  31.0   
7        2       Palasa             1 APPALARAJU SEEDIRI    MALE  39.0   
8        2       Palasa                2 GOUTHU SIREESHA  FEMALE  44.0   
9        2       Palasa         3 KOTHA PURNACHANDRA RAO    MALE  47.0   
10       2       Palasa                           4 NOTA     NaN   NaN   
11       2       Palasa            5 KORRAYI BALAKRISHNA    MALE  49.0   
12       2       Palasa                   6 MAJJI SARADA  FEMALE  64.0   
13       2       Palasa                 7 DUNNA VASUDEVU    MALE  54.0   
14       2       Palasa               8 MANOJ NALANAGULA    MALE  38.0   
15       2       Palasa     9 KAMESWARA RAO POTHANAPALLI    MALE  51.0   
16       2       Palasa                 10 MAMIDI PRASAD    MALE  32.0   
17       3      Tekkali          1 ATCHANNAIDU KINJARAPU    MALE  49.0   
18       3      Tekkali                   2 PERADA TILAK    MALE  45.0   
19       3      Tekkali                           3 NOTA     NaN   NaN   

   Category       Party  General  Postal  Total  % Votes Polled  Ranked1  \
0   GENERAL         TDP    79405   587.0  79992       46.392883      1.0   
1   GENERAL       YSRCP    71931   916.0  72847       42.249004      2.0   
2   GENERAL         JnP    10940   183.0  11123        6.450996      3.0   
3       NaN        NOTA     3872     8.0   3880        2.250280      4.0   
4   GENERAL         INC     2100    38.0   2138        1.239974      5.0   
5   GENERAL         BJP     1656   170.0   1826        1.059023      6.0   
6   GENERAL         IND      613     4.0    617        0.357841      7.0   
7   GENERAL       YSRCP    75357  1246.0  76603       50.834489      1.0   
8   GENERAL         TDP    59873   483.0  60356       40.052823      2.0   
9   GENERAL         JnP     5972   161.0   6133        4.069918      3.0   
10      NaN        NOTA     3034    10.0   3044        2.020028      4.0   
11  GENERAL         BJP     1359   215.0   1574        1.044522      5.0   
12  GENERAL         INC     1226    29.0   1255        0.832830      6.0   
13  GENERAL         IND      636     3.0    639        0.424047      7.0   
14  GENERAL         IND      392     NaN    392        0.260135      8.0   
15  GENERAL  CPI(ML)(L)      368     1.0    369        0.244872      9.0   
16  GENERAL         IND      322     4.0    326        0.216337     10.0   
17  GENERAL         TDP    86626  1032.0  87658       49.641245      1.0   
18  GENERAL       YSRCP    77769  1344.0  79113       44.802161      2.0   
19      NaN        NOTA     2929     6.0   2935        1.662108      3.0   

    Ranked  
0      1.0  
1      2.0  
2      3.0  
3      4.0  
4      5.0  
5      6.0  
6      7.0  
7      1.0  
8      2.0  
9      3.0  
10     4.0  
11     5.0  
12     6.0  
13     7.0  
14     8.0  
15     9.0  
16    10.0  
17     1.0  
18     2.0  
19     3.0

In [ ]:
nota = data.loc[data['Party']=='NOTA']
print(nota.groupby(['Party','Ranked']).size())



Party  Ranked
NOTA   3.0       19
       4.0       74
       4.5        1
       5.0       51
       6.0       25
       7.0        4
       9.0        1
dtype: int64


In [ ]:
cate= data.loc[data['Ranked']==1]
cate.groupby('Category').size()

Category
GENERAL    139
SC          28
ST           7
dtype: int64

In [ ]:
cate.groupby('Gender').size()

Gender
FEMALE     14
MALE      160
dtype: int64

In [ ]:
fav_party= data.loc[data['Party']=='TDP']

In [ ]:
inspect_cate =fav_party.loc[fav_party['Ranked']==1]

In [ ]:
inspect_cate.groupby('Category').size()

Category
GENERAL    22
SC          1
dtype: int64

In [ ]:
fav_party.groupby('Category').size()

Category
GENERAL    139
SC          29
ST           7
dtype: int64

In [ ]:
younge_candidates = data['Age']<30

In [ ]:
distribution =data[younge_candidates]

In [ ]:
distribution.groupby('Party').size()

Party
AIP            1
ANC            1
BJP            5
BSP            1
CPI(ML)(L)     2
DABAP          2
ILP(AP)        1
INC            2
IND           82
IPBP           2
IUML           4
JAJGP          3
JASWP          1
JHUP           1
JMMMP          2
JPD            2
JnP           10
MPP            5
NPRP           1
NPT            1
NRCP           2
NVSP           2
NavaP          1
PP             1
PPOI           8
PRSHP          5
SAPRP          1
SP             1
SUCI(C)        1
TDP            3
VCK            1
VMRMP          1
YSRCP          1
dtype: int64